### PR TITLE
feat(safetynet): simplify module entrypoint

### PR DESCRIPTION
**Pull request by ocardia**
## Description

Simplify the entrypoint of the Safetynet module. In this way, a new reader can immediately grasp the main logic without delving into details.

## Checklist

- [x] I have followed the indications in the [CONTRIBUTING](https://github.com/immuni-app/immuni-backend-analytics/blob/master/CONTRIBUTING.md).
- [x] The documentation related to the proposed change has been updated accordingly (plus comments in code).
- [ ] I have written new tests for my core changes, as applicable.
- [ ] I have successfully run tests with my changes locally.
- [x] It is ready for review! :rocket:

### CODE REVIEW

File: immuni_analytics/helpers/safety_net.py [hunk 1]
```diff
@@ -199,54 +259,10 @@ def verify_attestation(
...
-    except (JSONDecodeError, binascii.Error, IndexError, ValueError) as exc:
-        _LOGGER.warning(
-            "Could not retrieve payload from jws token.",
-            extra=dict(error=str(exc), safety_net_attestation=safety_net_attestation),
-        )
-        raise SafetyNetVerificationError()
-
+    header = _get_jws_header(safety_net_attestation)
+    certificates = _get_certificates(header)
+    leaf_certificate = _get_leaf_certificate_x509(certificates[0])
```
Comment (alessandroberlati) [inline immuni_analytics/helpers/safety_net.py]: I would pass the whole array here. The leaf can be retrieved by the logic

File: immuni_analytics/helpers/safety_net.py [hunk 2]
```diff
@@ -124,8 +178,14 @@ def _verify_signature(jws_token: str, leaf_certificate: crypto.x509) -> None:
...
-
-    jwt.decode(jws_token, public_key)
+    try:
+        jwt.decode(jws_token, public_key)
+    except DecodeError as exc:
+        _LOGGER.warning(
+            "Could not verify jws signature.",
+            extra=dict(error=str(exc), jws_token=jws_token, public_key=public_key),
+        )
+        raise SafetyNetVerificationError()
```
Comment (alessandroberlati) [inline immuni_analytics/helpers/safety_net.py:205]: Add raises for all method.

File: immuni_analytics/helpers/safety_net.py [hunk 3]
```diff
@@ -97,23 +105,69 @@ def _get_jws_payload(jws_token: str) -> Dict[str, Any]:
...
+            "Could not load the leaf certificate.",
+            extra=dict(error=str(exc), leaf_certificate_bytes=leaf_certificate_bytes),
+        )
+        raise SafetyNetVerificationError()
 
 
-def _verify_certificates(certificates: List) -> crypto.X509:
+def _validate_certificates(
+        leaf_certificate_x509: crypto.X509,
+        other_certificates_bytes: List[bytes]
```
Comment (alessandroberlati) [inline immuni_analytics/helpers/safety_net.py]: Not a fan of the Hungarian notation, we never used it in the whole codebase so I would avoid. What do you think?

File: immuni_analytics/helpers/safety_net.py [hunk 4]
```diff
@@ -97,23 +105,69 @@ def _get_jws_payload(jws_token: str) -> Dict[str, Any]:
...
+        _LOGGER.warning(
+            "Could not decode jws header certificates.",
+            extra=dict(error=str(exc), certificates_string=certificates_string),
+        )
+        raise SafetyNetVerificationError()
+
+    return certificates
+
+
+def _get_leaf_certificate_x509(leaf_certificate_bytes: bytes):
```
Comment (alessandroberlati) [inline immuni_analytics/helpers/safety_net.py]:
```suggestion
def _load_leaf_certificate_x509(leaf_certificate_bytes: bytes):
```

File: immuni_analytics/helpers/safety_net.py (hunk 3 above)
Comment (ocardia) [reply immuni_analytics/helpers/safety_net.py]: Fair enough, there's typing so it shouldn't be a concern!

File: immuni_analytics/helpers/safety_net.py (hunk 2 above)
Comment (ocardia) [reply immuni_analytics/helpers/safety_net.py:205]: Would public_key = crypto.dump_publickey(crypto.FILETYPE_PEM, leaf_certificate.get_pubkey()) raise a DecodeError?

File: immuni_analytics/helpers/safety_net.py (hunk 1 above)
Comment (ocardia) [reply immuni_analytics/helpers/safety_net.py]: IMHO, it made clearer that this method only needs the first, and the `other_certificates_bytes` in the subsequent method would get the rest.

Yet, do you agree that if I'm changing this method to get the whole certificates array, we should do the same in the call below too?

File: immuni_analytics/helpers/safety_net.py (hunk 2 above)
Comment (alessandroberlati) [reply immuni_analytics/helpers/safety_net.py:205]: No, I meant `raises` sorry. In the docstring for all _methods_ in the file :D

Comment (ocardia) [reply immuni_analytics/helpers/safety_net.py:205]: Are you the little picky one? 😝 Sure, I'll do ASAP. 🚀

File: immuni_analytics/helpers/safety_net.py (hunk 3 above)
Comment (ocardia) [reply immuni_analytics/helpers/safety_net.py]: Fixed ✅

File: immuni_analytics/helpers/safety_net.py (hunk 1 above)
Comment (ocardia) [reply immuni_analytics/helpers/safety_net.py]: Done ✅

File: immuni_analytics/helpers/safety_net.py (hunk 2 above)
Comment (ocardia) [reply immuni_analytics/helpers/safety_net.py:205]: Done. ✅

**Review by alessandroberlati — APPROVED** [review]

**Pull request merged**